ote_sdk/ote_sdk/entities/model.py

```diff
@@ -116,6 +124,49 @@ def __init__(
         model_size_reduction: float = 0.0,
         _id: Optional[ID] = None,
     ):
+        check_input_param_type(
+            RequiredParamTypeCheck(configuration, "configuration", ModelConfiguration),
+            OptionalParamTypeCheck(creation_date, "creation_date", datetime.datetime),
+            OptionalParamTypeCheck(performance, "performance", Performance),
+            OptionalParamTypeCheck(
+                previous_trained_revision, "previous_trained_revision", ModelEntity
+            ),
+            OptionalParamTypeCheck(previous_revision, "previous_revision", ModelEntity),
+            RequiredParamTypeCheck(version, "version", int),
+            OptionalParamTypeCheck(tags, "tags", List[str]),
+            RequiredParamTypeCheck(model_format, "model_format", ModelFormat),
+            RequiredParamTypeCheck(training_duration, "training_duration", float),
+            OptionalParamTypeCheck(
+                model_adapters, "model_adapters", Dict[str, ModelAdapter]
+            ),
+            OptionalParamTypeCheck(
+                exportable_code_adapter,
+                "exportable_code_adapter",
+                ExportableCodeAdapter,
+            ),
+            OptionalParamTypeCheck(precision, "precision", List[ModelPrecision]),
+            RequiredParamTypeCheck(latency, "latency", int),
+            RequiredParamTypeCheck(fps_throughput, "fps_throughput", int),
+            RequiredParamTypeCheck(target_device, "target_device", TargetDevice),
+            OptionalParamTypeCheck(target_device_type, "target_device_type", str),
+            RequiredParamTypeCheck(
+                optimization_type, "optimization_type", ModelOptimizationType
+            ),
+            OptionalParamTypeCheck(
+                optimization_methods, "optimization_methods", List[OptimizationMethod]
+            ),
+            OptionalParamTypeCheck(
+                optimization_objectives, "optimization_objectives", Dict[str, str]
+            ),
+            OptionalParamTypeCheck(
+                performance_improvement, "performance_improvement", Dict[str, float]
+            ),
+            RequiredParamTypeCheck(model_size_reduction, "model_size_reduction", float),
+            OptionalParamTypeCheck(_id, "_id", (ID, ObjectId)),
+        )
+        if train_dataset:
+            DatasetParamTypeCheck(train_dataset, "train_dataset").check()
```

Comment:
why is  `RequiredParamTypeCheck(train_dataset, "train_dataset", DatasetEntity)` not here?

Reply:
Because as the dataset-like parameter may be used non-child class of DatasetEntity, for example HpoDataset, which has same DatasetEntity attributes and methods, but not defined as child class

Reply:
It may mean that HpoDataset is wrong. @LeonidBeynenson what exact requirements on input parameters validation? I would not like to see in the SDK such a mess. In one place DatasetEntity is checked in one way, in another place it is checked in another way.

Reply:
I can add DatasetParamTypeCheck in all methods and functions, where dataset-like parameter is used

Reply:
Added DatasetParamTypeCheck in all methods and functions, where dataset-like parameter is used

Reply:
Thanks, but it is better to have InputParamTypeCheck for everything.

Reply:
Ok, lets wait for exact requirements from Leonid

Reply:
Could you investigate whether method can be decorated so that there is no need to list parameters, their names, their types, for example https://stackoverflow.com/questions/15299878/how-to-use-python-decorators-to-check-function-arguments

Reply:
Thanks, added draft version of check_input_parameters_type decorator and updated check_parameter_type function

Reply:
Thanks, added “check_input_parameters_type” decorator, since there are custom checks required, added checks_types parameter dictionary with custom checks